inc/namespace.php

```diff
@@ -567,44 +567,51 @@ function set_tachyon_hostname( string $tachyon_url ) : string {
 
-	if ( ! $primary_host ) {
-		trigger_error( sprintf( 'Error parsing main site URL: %s', esc_url_raw( get_main_site_url() ) ), E_USER_WARNING );
-		return $dirs;
-	}
+	if ( ! $s3_upload_dirs ) {
+		$primary_host = wp_parse_url( get_main_site_url(), PHP_URL_HOST );
+		$s3_host = wp_parse_url( $s3_uploads->get_s3_url(), PHP_URL_HOST );
+		$current_host = wp_parse_url( site_url(), PHP_URL_HOST );
 
-	if ( ! $s3_host ) {
-		trigger_error( sprintf( 'Error parsing S3 bucket URL: %s', esc_url_raw( $s3_uploads->get_s3_url() ) ), E_USER_WARNING );
-		return $dirs;
-	}
+		if ( ! $primary_host ) {
+			trigger_error( sprintf( 'Error parsing main site URL: %s', esc_url_raw( get_main_site_url() ) ), E_USER_WARNING );
+			return $dirs;
+		}
 
-	if ( ! $current_host ) {
-		trigger_error( sprintf( 'Error parsing site URL: %s', esc_url_raw( site_url() ) ), E_USER_WARNING );
-		return $dirs;
-	}
+		if ( ! $s3_host ) {
+			trigger_error( sprintf( 'Error parsing S3 bucket URL: %s', esc_url_raw( $s3_uploads->get_s3_url() ) ), E_USER_WARNING );
+			return $dirs;
+		}
 
-	// To support 3rd party CDNs leave the host names as is if the direct
-	// amazonaws.com URL is in use.
-	if ( strpos( $s3_host, '.amazonaws.com' ) !== false ) {
-		return $dirs;
-	}
+		if ( ! $current_host ) {
+			trigger_error( sprintf( 'Error parsing site URL: %s', esc_url_raw( site_url() ) ), E_USER_WARNING );
+			return $dirs;
+		}
 
-	// Ensure uploads host at least matches primary site host.
-	if ( $s3_host !== $primary_host ) {
-		$dirs['url'] = str_replace( "://{$s3_host}", "://{$primary_host}", $dirs['url'] );
-		$dirs['baseurl'] = str_replace( "://{$s3_host}", "://{$primary_host}", $dirs['baseurl'] );
-	}
+		// To support 3rd party CDNs leave the host names as is if the direct
+		// amazonaws.com URL is in use.
+		if ( strpos( $s3_host, '.amazonaws.com' ) !== false ) {
+			return $dirs;
+		}
+
+		// Ensure uploads host at least matches primary site host.
+		if ( $s3_host !== $primary_host ) {
+			$dirs['url'] = str_replace( "://{$s3_host}", "://{$primary_host}", $dirs['url'] );
+			$dirs['baseurl'] = str_replace( "://{$s3_host}", "://{$primary_host}", $dirs['baseurl'] );
+		}
+
+		// Only do the replacement if the host name is not a subdomain of the S3 host.
+		if ( substr( $current_host, -1 * strlen( $primary_host ) ) !== $primary_host ) {
+			$dirs['url'] = str_replace( "://{$primary_host}", "://{$current_host}", $dirs['url'] );
+			$dirs['baseurl'] = str_replace( "://{$primary_host}", "://{$current_host}", $dirs['baseurl'] );
+		}
 
-	// Only do the replacement if the host name is not a subdomain of the S3 host.
-	if ( substr( $current_host, -1 * strlen( $primary_host ) ) !== $primary_host ) {
-		$dirs['url'] = str_replace( "://{$primary_host}", "://{$current_host}", $dirs['url'] );
-		$dirs['baseurl'] = str_replace( "://{$primary_host}", "://{$current_host}", $dirs['baseurl'] );
+		$s3_upload_dirs = $dirs;
+		wp_cache_set( 's3_upload_dirs', $s3_upload_dirs );
```

Comment:
Pretty sure this will need to be keyed or grouped on at least the `$current_host` as it's dynamic for each site. Possibly on the S3 host too but pretty sure that's static.

Reply:
Yeah, that's one thing I wasn't sure about. I thought that the object cache would automatically key entries by the site ID, but I should really validate that assumption.

In any case here, this isn't even a persistant entry so there _shouldn't_ be any way for it to need further grouping?

Reply:
As I understand it any call to the object cache will use redis which is persistent. Maybe I’m missing something though

Reply:
:facepalm: Nope, that's my bad. For asome reason I had it in my mind that not specifying an expiration meant creating a non-persistant cache entry. In actuality its just the opposite - no expiration means eternal.

Reply:
Also fwiw I was wrong - caches are keyed for the current site unless a group is specified that's registered as a global cache group so using `wp_cache_set()` is fine.

Reply:
I'm happy to stick with using a static variable though for now - there are a few possibilities I could think of where caching this value between requests might cause issues (the site URL could change, one site could serve requests for more than one hostname, http vs https, etc.), and in any case the problem I was trying to guard against was just that this happen multiple times in one request, not that it happens at all.

Reply:
Makes sense to me 👍  Lower risk this way too.